docs-issue-generation: Script can now handle multiple release notes

nickvigilante opened the pull request:
Previously, if a commit to the cockroach repo contained multiple
RNs, the docs issue generation script only evaluated the
last release note. Additionally, if the release justification was
included below the last release note in a commit, it would be
included as part of the release note. Now, each release note in a
commit is evaluated, and the release justification is stripped.

Fixes https://github.com/cockroachdb/cockroach/issues/84434

Release note: None

Release justification: non-production code change

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/85490)
<!-- Reviewable:end -->

jlinder reviewed (changes requested):
Cool. The core intention of what you want the change to do looks good. Here are some comments for improvements.

> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-:-N9K0XII7UMuSQkJ_itc:b-yc7ppm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nickvigilante)

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9DUTCu34v1Y9lsYCtl):*
A couple comments on this line. They also apply to the PR title and body:

- In the cockroach repo, the first line format is "area-of-code: message". 
- The reference to the issue the commit is related to should be placed in the body of the message. It should use one of [the standard GitHub issue reference verbs](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) or one of `See also:` or `Informs:`.
- When referencing a related issue, prefer the GitHub version of the issue if there is one. For this commit, it should be referencing https://github.com/cockroachdb/cockroach/issues/84434 .
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9IfhRs9schNhMWx-E_:-N9IfhRu-i1bMNRmdxaR:b-a1qsi1) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c5aa82b73b2d74327f065e482961bc61e9c812c/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L28)):*
> ```Go
> // commit contains details about each formatted commit
> type issue struct {
> 	sha         string
> ```

Suggestion:
```suggestion
	sourceCommitSha         string
```
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9DieRQ7PB9o_gREuTY:-N9DieRQ7PB9o_gREuTZ:b-enh06z) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c5aa82b73b2d74327f065e482961bc61e9c812c/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L30)):*
> ```Go
> 	sha         string
> 	title       string
> 	releaseNote string
> ```

Suggestion:
```suggestion
	body string
```

Suggesting this because `issue` is in the name of the struct. With this change, referencing the field is `docsIssue.body` instead of `docsIssue.issueBody`. This is a example of following the [DRY principle](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself).

You might also apply this logic to `issueTitle` above.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 135 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9IgaORBEJyhVG6gOzx:-N9IgaORBEJyhVG6gOzy:b46uzyi) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c5aa82b73b2d74327f065e482961bc61e9c812c/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L135)):*
> ```Go
> }
> 
> // getIssues takes a list of commits from GitHub as well as the PR associated with those commits and outputs a formatted list of commits with valid release notes on that PR
> ```

This comment also needs to be updated to reflect the new output of the function. 

It's also a really long comment. Maybe limit its length to a shorter length? When lines are this long it makes it harder to work on just a laptop screen or have multiple columns of code next to each other on the same screen. It also makes side-by-side code review harder when long lines wrap and break the visual indentation flow.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 26 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9Ilxmn21osLAtpDfmQ:-N9Ilxmn21osLAtpDfmR:brw93wq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L26)):*
> ```Go
> )
> 
> // docsIssue contains details about each formatted commit to be committed to the docs repo
> ```

Suggestion:
```suggestion
// docsIssue contains details about each formatted commit to be committed to the docs repo
```
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 41 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9InV5a34HsjlBst3Bn:-N9InV5a34HsjlBst3Bo:b-l384et) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L41)):*
> ```Go
> 
> // ghSearch contains a list of search items (in this case, cockroach PRs) from the GitHub API.
> // This struct holds the search results of all PRs based on a given commit.
> ```

Is this line redundant to the line above it? Mabye combine them somehow?
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 49 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9Ini2LC9Vbrsh7iMJf:-N9Ini2LC9Vbrsh7iMJg:b-w23uu9) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L49)):*
> ```Go
> // including its number, from the GitHub API.
> type ghSearchItem struct {
> 	Number      int            `json:"number"` // PR number
> ```

Maybe a better name for the field is `PRNumber`? 
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 58 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9Io8CEDg9Lc78CNtzP:-N9Io8CEDg9Lc78CNtzQ:b5w6ww3) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L58)):*
> ```Go
> }
> 
> // this struct holds the base branch for a cockroach PR from the GitHub API
> ```

Could you change this comment to follow the standard [go doc commenting format](https://tip.golang.org/doc/comment) for types?
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 67 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9Ip-EK33kRr_ivFsjd:-N9Ip-EK33kRr_ivFsje:b-vodk78) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L67)):*
> ```Go
> // ghPullCommit holds the SHA and commit message for a particular commit, from the GitHub API
> type ghPullCommit struct {
> 	CRDBSha       string          `json:"crdbSha"`
> ```

Why does this change need to happen?

The JSON coming in from GitHub will use the field name `sha`, so the annotation `json:"sha"` needs to stay as it was. 

And the variable name doesn't need to explicitly say `CRDB`. The data this struct will hold could be from any PR (it doesn't logically need to be from a cockroach PR). The context of how this is used doesn't need to be explicit that it's a cockroach sha.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 68 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9Iq7EI4mVKXfbfrV-w:-N9Iq7EI4mVKXfbfrV-x:bgblg18) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L68)):*
> ```Go
> type ghPullCommit struct {
> 	CRDBSha       string          `json:"crdbSha"`
> 	CRDBCommitMsg ghPullCommitMsg `json:"commit"`
> ```

Ditto re the name of the variable not needing to change and already being descriptive enough.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 82 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9IpQmJ4a9iwEhdr7ly:-N9IpQmJ4a9iwEhdr7lz:b-b6y4tq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L82)):*
> ```Go
> }
> 
> // the heart of the script to fetch and manipulate all data and create the individual docs docsIssues
> ```

There are a lot of places in the comments and strings where `issue` changed to `docsIssue` and it doesn't seem like what should have happened (it's like a replace all was run and it found places that shouldn't have changed). Can you double check this was meant everywhere?
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 142 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9K1RYrEzXGqJvUPfsx:-N9K1RYrEzXGqJvUPfsy:b4yxtm1) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L142)):*
> ```Go
> 	for _, c := range pullCommit {
> 		message := c.CRDBCommitMsg.Message
> 		rns := formatReleaseNotes(message, prNumber, c.CRDBSha) // return a slice of data that locates every instance of the phrase "release note (", case insensitive
> ```

Could the end-of-line comments be moved before the lines they are for in this function too?
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 158 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9JjShXAEHbbKRpEoTg:-N9JjShYCXJEF0HsheI5:bh8afu9) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L158)):*
> ```Go
> 
> func formatReleaseNotes(message string, prNumber int, crdbSha string) []string {
> 	// formatReleaseNotes generates a list of docsIssue bodies for the docs repo
> ```

This comment should be before the function as the function doc comment. It's describing what the function does.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 161 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9JlYR37cdUsatYOREr:-N9JlYR37cdUsatYOREs:b-tu48sz) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L161)):*
> ```Go
> 	// based on a given crdbSha
> 	rnBodySlice := []string{}
> 	noRN := regexp.MustCompile(`[rR]elease [nN]ote: [nN]one`)
> ```

Since the regexs compiled in this function don't change and this function is called many times, it would be good to move them out to be vars in the module like in [this example](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/smithtest/main.go#L132-L136). Then they will be compiled only once and used many times.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 162 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9JjxwT0yQpYd5vatty:-N9JjxwT0yQpYd5vattz:b81h2gd) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L162)):*
> ```Go
> 	rnBodySlice := []string{}
> 	noRN := regexp.MustCompile(`[rR]elease [nN]ote: [nN]one`)
> 	if !noRN.MatchString(message) {
> ```

How about inverting the logic in this `if` statement and then out-denting the body of it? This will keep the level of indents in the function lower and make readability better, especially when there is a long body to the block and there is no `else` at the end.

```suggestion
	if noRN.MatchString(message) {
		return rnBodySlice
	}
```
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 167 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9JomhJ3vxsM7gFKONH:-N9JomhN48bXoD9ER0Ew:bops3qu) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L167)):*
> ```Go
> 		bugFixRNRE := regexp.MustCompile(`([rR]elease [nN]ote \(bug fix\):.*)`)
> 		releaseJustificationRE := regexp.MustCompile(`[rR]elease [jJ]ustification:.*`)
> 		buffer := ""
> ```

Two things here:

1) Maybe name this `releaseNoteLines` (or similar) to indicate what the variable contains?

2) I recommend building the release note string using a slice instead of successively concatenating strings together. See method 3 on [this page](https://golangdocs.com/concatenate-strings-in-golang).

When using the `+` operator to join two strings, an entirely new string is created for every `+` operator that gets processed (which means the contents of the two strings are copied into a new memory location for the new string). For a 10 line release note, this means it would create 20 different strings (because of the newlines being appended too) with each one being longer than the previous contents of the variable.

Using a slice to concatenate many strings together allows for an efficient creation of the final string by only copying each source string to the final string once.

This problem and solution apply to many (most?) other languages too (JavaScript, Python, Ruby, Java, ...).
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 174 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9Jt6bkDg4JMEjVWlmr:-N9Jt6bm-DDbIdFWP2_s:b-5yfw2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L174)):*
> ```Go
> 			releaseJustification := releaseJustificationRE.MatchString(x)
> 			if bufferNotEmpty && (validRn || releaseJustification) {
> 				rnBodySlice = append(rnBodySlice, strings.TrimSuffix(fmt.Sprintf(
> ```

I'd recommend generation of the new release note body into a variable before adding it to the `rnBodySlice`. That will make what is happening much more clear and the indentation will be easier to handle.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 175 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9JtrJuChL2kHw3wKwG:-N9JtrJuChL2kHw3wKwH:b-m41hxs) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L175)):*
> ```Go
> 			if bufferNotEmpty && (validRn || releaseJustification) {
> 				rnBodySlice = append(rnBodySlice, strings.TrimSuffix(fmt.Sprintf(
> 					"Related PR: https://github.com/cockroachdb/cockroach/pull/%s\nCommit: https://github.com/cockroachdb/cockroach/commit/%s\n\n---\n\n%s",
> ```

Suggestion:
```suggestion
					"Related PR: https://github.com/cockroachdb/cockroach/pull/%s\n" +
					"Commit: https://github.com/cockroachdb/cockroach/commit/%s\n" +
					"\n---\n\n%s",
```

This will make it much easier to understand what the resulting text will look like when the string is rendered in the issue.

Also, using the `+` operator in cases where there are a harde-coded strings together like this is OK. It's just when the string being generated will be of unknown length with lots of concatenations that it can become a performance problem.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 190 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9JvU6v4zx301OBYVEg:-N9JvU6v4zx301OBYVEh:b-wfh5ww) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L190)):*
> ```Go
> 		}
> 		if buffer != "" { // commit whatever is left in the buffer to the rnBodySlice set
> 			rnBodySlice = append(rnBodySlice, strings.TrimSuffix(fmt.Sprintf(
> ```

Ditto re: generating the new release note body into a variable before adding it to the slice.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 211 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9JvdLf9XEqTd-tfYCz:-N9JvdLf9XEqTd-tfYD-:bajmduz) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L211)):*
> ```Go
> 		result = message
> 	}
> 	result = fmt.Sprintf("PR #%d - %s",
> ```

Two things:

1. It looks like `result` will always be overwritten here. So whatever was set in the `if...else` above will be lost.
2. Statements like this are usually be done on one line. It's short and clear what is happening.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 229 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9JwkHXEt0JyvpN9K2w:-N9JwkHXEt0JyvpN9K2x:b2zrzxw) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L229)):*
> ```Go
> 	for _, issue := range p.docsIssues {
> 		reqBody, err := json.Marshal(map[string]interface{}{
> 			"issueTitle": issue.issueTitle,
> ```

This should be:

```
			"title": issue.issueTitle,
```
___
*[`pkg/cmd/docs-issue-generation/main.go` line 12 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9K5HEX2VxU-2r6y4XQ:-N9K5HEZ0go7c1wMRq_S:b5qhpcx) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/main.go#L12)):*
> ```Go
> 
> // Check that GitHub PR descriptions and commit messages contain the
> // expected epic and docsIssue references.
> ```

This file also has some strings that look like they've been updated by mistake.


<!-- Sent from Reviewable.io -->

nickvigilante reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85490)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jlinder and @nickvigilante)

___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9IfhRs9schNhMWx-E_:-N9gKugnAWMzUbOqtE5v:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c5aa82b73b2d74327f065e482961bc61e9c812c/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L28)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Suggestion:
```suggestion
	sourceCommitSha         string
```


</blockquote></details>

Done.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9DieRQ7PB9o_gREuTY:-N9gL5Y99tO732pUzV6s:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c5aa82b73b2d74327f065e482961bc61e9c812c/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L30)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Suggestion:
```suggestion
	body string
```

Suggesting this because `issue` is in the name of the struct. With this change, referencing the field is `docsIssue.body` instead of `docsIssue.issueBody`. This is a example of following the [DRY principle](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself).

You might also apply this logic to `issueTitle` above.
</blockquote></details>

Done.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 135 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9IgaORBEJyhVG6gOzx:-N9hcgBT3u95Q7DIa660:bfjnij8) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c5aa82b73b2d74327f065e482961bc61e9c812c/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L135)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

This comment also needs to be updated to reflect the new output of the function. 

It's also a really long comment. Maybe limit its length to a shorter length? When lines are this long it makes it harder to work on just a laptop screen or have multiple columns of code next to each other on the same screen. It also makes side-by-side code review harder when long lines wrap and break the visual indentation flow.
</blockquote></details>

Understood. I've split the comment across multiple lines. Also, I moved some of the function calls within this file to have their parameters exist on their own lines.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 26 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9Ilxmn21osLAtpDfmQ:-N9gKjjbCwYPsIiIyB9D:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L26)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Suggestion:
```suggestion
// docsIssue contains details about each formatted commit to be committed to the docs repo
```
</blockquote></details>

Done.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 41 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9InV5a34HsjlBst3Bn:-N9gLGKZ5DoHS3seLYir:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L41)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Is this line redundant to the line above it? Mabye combine them somehow?
</blockquote></details>

Done.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 49 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9Ini2LC9Vbrsh7iMJf:-N9gLODG8hIbGZ1gqb4Y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L49)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Maybe a better name for the field is `PRNumber`? 
</blockquote></details>

Done.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 58 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9Io8CEDg9Lc78CNtzP:-N9gNNxxBOuzclKXP8hO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L58)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Could you change this comment to follow the standard [go doc commenting format](https://tip.golang.org/doc/comment) for types?
</blockquote></details>

Done.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 67 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9Ip-EK33kRr_ivFsjd:-N9gNp5n-v4ynn86sfYe:b9buvsp) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L67)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Why does this change need to happen?

The JSON coming in from GitHub will use the field name `sha`, so the annotation `json:"sha"` needs to stay as it was. 

And the variable name doesn't need to explicitly say `CRDB`. The data this struct will hold could be from any PR (it doesn't logically need to be from a cockroach PR). The context of how this is used doesn't need to be explicit that it's a cockroach sha.

</blockquote></details>

Accidentally got renamed during my refactor in Goland. Reverted/done.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 68 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9Iq7EI4mVKXfbfrV-w:-N9gOMrN-R_kOvRg2h5o:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L68)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Ditto re the name of the variable not needing to change and already being descriptive enough.
</blockquote></details>

Done.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 82 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9IpQmJ4a9iwEhdr7ly:-N9gOwnLCrIGV9HpX-ZP:b-m12g4s) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L82)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

There are a lot of places in the comments and strings where `issue` changed to `docsIssue` and it doesn't seem like what should have happened (it's like a replace all was run and it found places that shouldn't have changed). Can you double check this was meant everywhere?
</blockquote></details>

Caused by the refactor in Goland. Fixed.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 142 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9K1RYrEzXGqJvUPfsx:-N9hdIGyEz_r48ga2aKG:bxyvo9t) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L142)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Could the end-of-line comments be moved before the lines they are for in this function too?
</blockquote></details>

Yes. I did this in other places that had lines longer than 115 characters.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 158 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9JjShXAEHbbKRpEoTg:-N9hdxnGANPru7eWYI9D:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L158)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

This comment should be before the function as the function doc comment. It's describing what the function does.
</blockquote></details>

Done.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 161 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9JlYR37cdUsatYOREr:-N9hea_QE5VdTCf4BG_9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L161)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Since the regexs compiled in this function don't change and this function is called many times, it would be good to move them out to be vars in the module like in [this example](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/smithtest/main.go#L132-L136). Then they will be compiled only once and used many times.
</blockquote></details>

Done.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 162 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9JjxwT0yQpYd5vatty:-N9hejcz7QAfT-Rg9TgN:brzmhyq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L162)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

How about inverting the logic in this `if` statement and then out-denting the body of it? This will keep the level of indents in the function lower and make readability better, especially when there is a long body to the block and there is no `else` at the end.

```suggestion
	if noRN.MatchString(message) {
		return rnBodySlice
	}
```

</blockquote></details>

I think I understand what you're asking here, and I agree. Hoping that this takes care of it.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 167 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9JomhJ3vxsM7gFKONH:-N9hgc8C8YD0VD8BP7gr:bx14lqd) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L167)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Two things here:

1) Maybe name this `releaseNoteLines` (or similar) to indicate what the variable contains?

2) I recommend building the release note string using a slice instead of successively concatenating strings together. See method 3 on [this page](https://golangdocs.com/concatenate-strings-in-golang).

When using the `+` operator to join two strings, an entirely new string is created for every `+` operator that gets processed (which means the contents of the two strings are copied into a new memory location for the new string). For a 10 line release note, this means it would create 20 different strings (because of the newlines being appended too) with each one being longer than the previous contents of the variable.

Using a slice to concatenate many strings together allows for an efficient creation of the final string by only copying each source string to the final string once.

This problem and solution apply to many (most?) other languages too (JavaScript, Python, Ruby, Java, ...).
</blockquote></details>

`releaseNoteLines` is now a string slice. I also use `append()` to add a new line to `releaseNoteLines`. I hope that's okay.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 174 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9Jt6bkDg4JMEjVWlmr:-N9hhab_9WmdPaCT-wl5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L174)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

I'd recommend generation of the new release note body into a variable before adding it to the `rnBodySlice`. That will make what is happening much more clear and the indentation will be easier to handle.
</blockquote></details>

Done.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 175 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9JtrJuChL2kHw3wKwG:-N9hhqaP-MpXDJ-tMROH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L175)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Suggestion:
```suggestion
					"Related PR: https://github.com/cockroachdb/cockroach/pull/%s\n" +
					"Commit: https://github.com/cockroachdb/cockroach/commit/%s\n" +
					"\n---\n\n%s",
```

This will make it much easier to understand what the resulting text will look like when the string is rendered in the issue.

Also, using the `+` operator in cases where there are a harde-coded strings together like this is OK. It's just when the string being generated will be of unknown length with lots of concatenations that it can become a performance problem.
</blockquote></details>

Done.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 190 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9JvU6v4zx301OBYVEg:-N9hhrby3TXlMkfCp7-f:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L190)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Ditto re: generating the new release note body into a variable before adding it to the slice.
</blockquote></details>

Done.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 211 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9JvdLf9XEqTd-tfYCz:-N9hiogj40E25V62rQYi:b-d2jgr3) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L211)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Two things:

1. It looks like `result` will always be overwritten here. So whatever was set in the `if...else` above will be lost.
2. Statements like this are usually be done on one line. It's short and clear what is happening.
</blockquote></details>

I added an intermediate variable called `commitTitle` to hold the commit title here.

Is it an antipattern if I overwrite a variable with an expression containing itself?
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 229 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9JwkHXEt0JyvpN9K2w:-N9hi1ig4DJKUmjqOZzx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L229)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

This should be:

```
			"title": issue.issueTitle,
```


</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

nickvigilante reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85490)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jlinder and @nickvigilante)

___
*[`pkg/cmd/docs-issue-generation/main.go` line 12 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9K5HEX2VxU-2r6y4XQ:-N9hjgvJ7jQAKcaq9eR_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/main.go#L12)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

This file also has some strings that look like they've been updated by mistake.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

Commit:
docs-issue-generation: Script can now handle multiple release notes

Previously, if a commit to the cockroach repo contained multiple
RNs, the docs issue generation script only evaluated the
last release note. Additionally, if the release justification was
included below the last release note in a commit, it would be
included as part of the release note. Now, each release note in a
commit is evaluated, and the release justification is stripped.

Release note: None

Release justification: non-production code change

jlinder reviewed (changes requested):
Great! I've a few more comments. 

> Reviewed 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-:-NAr58TW38bFTWgvo5YX:b-96akw2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nickvigilante)

___
*[`-- commits` line 14 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-NAqmYUh7fCKziq_9cvh):*
The release justification can just be:

```suggestion
Release justification: non-production code change
```
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 135 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9IgaORBEJyhVG6gOzx:-NAqqxyL940ILhoI89zv:b1wvabo) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c5aa82b73b2d74327f065e482961bc61e9c812c/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L135)):*
<details><summary><i>Previously, nickvigilante (Nick Vigilante) wrote…</i></summary><blockquote>

Understood. I've split the comment across multiple lines. Also, I moved some of the function calls within this file to have their parameters exist on their own lines.
</blockquote></details>

Cool. :thumbsup: to the formatting change for long function calls. That makes them much easier to understand.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 162 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9JjxwT0yQpYd5vatty:-NAqsOBq-1K1Iwcnh4cq:bnp63wg) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L162)):*
<details><summary><i>Previously, nickvigilante (Nick Vigilante) wrote…</i></summary><blockquote>

I think I understand what you're asking here, and I agree. Hoping that this takes care of it.
</blockquote></details>

Excellent. Yes, you got what I was asking. Thanks!
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 167 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9JomhJ3vxsM7gFKONH:-NAr-Km42llu2HIuhVo2:b-ryhrvp) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L167)):*
<details><summary><i>Previously, nickvigilante (Nick Vigilante) wrote…</i></summary><blockquote>

`releaseNoteLines` is now a string slice. I also use `append()` to add a new line to `releaseNoteLines`. I hope that's okay.
</blockquote></details>

Great! Yes, using `append()` here is appropriate and expected. 
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 211 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9JvdLf9XEqTd-tfYCz:-NAr-fi2-vlp2UuTaHHk:b-hixpae) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L211)):*
<details><summary><i>Previously, nickvigilante (Nick Vigilante) wrote…</i></summary><blockquote>

I added an intermediate variable called `commitTitle` to hold the commit title here.

Is it an antipattern if I overwrite a variable with an expression containing itself?
</blockquote></details>

Ah! I think I missed that `result` was used in the `fmt.Sprintf` call. 

There are certainly times when overwriting the variable like you were doing with `result` is appropriate. In this case, using `commitTitle` as the initial variable makes this code much clearer because it `commitTitle` variable now states specifically what it represents. If it had that name originally, I don't think I would have left the comment.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 91 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-NAqoiGAD2gs8rbz5FW9:-NAqoiGAD2gs8rbz5FWA:b-r63mm8) ([raw file](https://github.com/cockroachdb/cockroach/blob/d21667e55805dc527991feb248bae359741caca1/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L91)):*
> ```Go
> 		params.Token,
> 		&search,
> 	); err != nil {
> ```

I'd suggest separating out the call to `httpGet()` into a separate line like this:

```suggestion
	err := httpGet(
		"https://api.github.com/search/issues?q=sha:"+params.Sha+"+repo:cockroachdb/cockroach+is:merged",
		params.Token,
		&search,
	)
	if err != nil {
```

Putting the very long call to `httpGet()` within the `if` statement makes it harder to understand the if statement. And since most people who interact with this code will be reading it and trying to understand it after this PR is merged, it's much better to keep the code simpler to understand.



_Code quote:_
```Go
	if err := httpGet(
		"https://api.github.com/search/issues?q=sha:"+params.Sha+"+repo:cockroachdb/cockroach+is:merged",
		params.Token,
		&search,
	); err != nil {
```
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 97 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-NAqq3j395LisNM7r5t6:-NAqq3j395LisNM7r5t7:b9gi9it) ([raw file](https://github.com/cockroachdb/cockroach/blob/d21667e55805dc527991feb248bae359741caca1/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L97)):*
> ```Go
> 	for _, pr := range prs { // for each PR in the list, get the merge branch and the list of eligible commits
> 		var pull ghPull
> 		if err := httpGet(
> ```

Ditto here on separating out the call to `httpGet()` into a separate line.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 106 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-NAqq80J4porBPT_s_CP:-NAqq80J4porBPT_s_CQ:b-rve1uq) ([raw file](https://github.com/cockroachdb/cockroach/blob/d21667e55805dc527991feb248bae359741caca1/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L106)):*
> ```Go
> 		pr.mergeBranch = pull.Base.Ref
> 		var commits []ghPullCommit
> 		if err := httpGet(
> ```

And ditto here on separating out the call to `httpGet()` into a separate line.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 176 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-NAqrhqb5v4A_J15dpTk:-NAqrhqb5v4A_J15dpTl:b7fqp8m) ([raw file](https://github.com/cockroachdb/cockroach/blob/d21667e55805dc527991feb248bae359741caca1/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L176)):*
> ```Go
> 
> var (
> 	noRN                   = regexp.MustCompile(`[rR]elease [nN]ote: [nN]one`)
> ```

Since the other variable names for regexps all end with `RE`, how about making this one consistent with them?

```suggestion
	noRNRE                 = regexp.MustCompile(`[rR]elease [nN]ote: [nN]one`)
```
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 193 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-NAqtjqbDwjGo2Lx9w-R:-NAqtjqbDwjGo2Lx9w-S:bj1vwto) ([raw file](https://github.com/cockroachdb/cockroach/blob/d21667e55805dc527991feb248bae359741caca1/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L193)):*
I'd suggest inserting `len(releaseNoteLines) > 0` directly into the `if` statements where `bufferNotEmpty` is used. The biggest reason is because possible side effects if the contents of `releaseNoteLines` change between the two uses of `bufferNotEmpty`. Using a pre-calculated value based on a variable that can change between the uses of the pre-calculated value in two `if` statements that are somewhat complex on their own is complicated. It's better to simplify and directly code in the calculation in both places.

I go back to these two ideas from the [Zen of Python](https://peps.python.org/pep-0020/) all the time. I think they apply here well:

> Simple is better than complex.
> Complex is better than complicated.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 198 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-NAqzCCGEvlW2Aena4s3:-NAqzCCH7upZGyxZxZfh:b-jke7sg) ([raw file](https://github.com/cockroachdb/cockroach/blob/d21667e55805dc527991feb248bae359741caca1/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L198)):*
> ```Go
> 		// if the buffer isn't empty and the next line is either a release justification or a new RN, write the buffer
> 		// to rnBodySlice and flush the buffer
> 		if bufferNotEmpty && (validRn || releaseJustification) {
> ```

Does the comment for this `if` block state the same thing as the code itself states? If the words in the comment are more descriptive than the variable names, would it be better to change the names of the variables?

If you do choose to keep the comment, maybe update it to use language related to "release note lines" instead of the old "buffer" related language?



_Code quote:_
```Go
		// if the buffer isn't empty and the next line is either a release justification or a new RN, write the buffer
		// to rnBodySlice and flush the buffer
		if bufferNotEmpty && (validRn || releaseJustification) {
```
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 244 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-NAr0bwP5FpZ6ucl1Zyv:-NAr0bwP5FpZ6ucl1Zyw:by9u0ud) ([raw file](https://github.com/cockroachdb/cockroach/blob/d21667e55805dc527991feb248bae359741caca1/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L244)):*
> ```Go
> 
> func (p pr) createDocsIssues(token string) {
> 	postURL := "https://api.github.com/repos/cockroachdb/docs/docsIssues"
> ```

It looks like this URL was hit by the search and replace too?

```suggestion
	postURL := "https://api.github.com/repos/cockroachdb/docs/issues"
 ```


<!-- Sent from Reviewable.io -->

nickvigilante reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85490)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jlinder)

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9DUTCu34v1Y9lsYCtl):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

A couple comments on this line. They also apply to the PR title and body:

- In the cockroach repo, the first line format is "area-of-code: message". 
- The reference to the issue the commit is related to should be placed in the body of the message. It should use one of [the standard GitHub issue reference verbs](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) or one of `See also:` or `Informs:`.
- When referencing a related issue, prefer the GitHub version of the issue if there is one. For this commit, it should be referencing https://github.com/cockroachdb/cockroach/issues/84434 .

</blockquote></details>

Done.
___
*[`-- commits` line 14 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-NAqmYUh7fCKziq_9cvh):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

The release justification can just be:

```suggestion
Release justification: non-production code change
```
</blockquote></details>

Done.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 135 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9IgaORBEJyhVG6gOzx:-NAtLyEHAEkBN4SWxriD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c5aa82b73b2d74327f065e482961bc61e9c812c/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L135)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Cool. :thumbsup: to the formatting change for long function calls. That makes them much easier to understand.
</blockquote></details>

Done.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 162 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9JjxwT0yQpYd5vatty:-NAtMpTa35TBjHN4CjFX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L162)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Excellent. Yes, you got what I was asking. Thanks!
</blockquote></details>

Done.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 167 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9JomhJ3vxsM7gFKONH:-NAtMsU6AdhTTl_goMM8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L167)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Great! Yes, using `append()` here is appropriate and expected. 
</blockquote></details>

Done.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 211 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-N9JvdLf9XEqTd-tfYCz:-NAtN43HAP9XklIN6kFT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432640543631d352c9e7e2fe157cdb6255f40a4/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L211)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Ah! I think I missed that `result` was used in the `fmt.Sprintf` call. 

There are certainly times when overwriting the variable like you were doing with `result` is appropriate. In this case, using `commitTitle` as the initial variable makes this code much clearer because it `commitTitle` variable now states specifically what it represents. If it had that name originally, I don't think I would have left the comment.
</blockquote></details>

Done.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 91 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-NAqoiGAD2gs8rbz5FW9:-NAtPJNd9AaEA0BruCHL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d21667e55805dc527991feb248bae359741caca1/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L91)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

I'd suggest separating out the call to `httpGet()` into a separate line like this:

```suggestion
	err := httpGet(
		"https://api.github.com/search/issues?q=sha:"+params.Sha+"+repo:cockroachdb/cockroach+is:merged",
		params.Token,
		&search,
	)
	if err != nil {
```

Putting the very long call to `httpGet()` within the `if` statement makes it harder to understand the if statement. And since most people who interact with this code will be reading it and trying to understand it after this PR is merged, it's much better to keep the code simpler to understand.
</blockquote></details>

Done.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 97 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-NAqq3j395LisNM7r5t6:-NAtQWqFDxSD9ymL0FrG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d21667e55805dc527991feb248bae359741caca1/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L97)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Ditto here on separating out the call to `httpGet()` into a separate line.
</blockquote></details>

Done.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 106 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-NAqq80J4porBPT_s_CP:-NAtQXuD-J6qne3uy3hv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d21667e55805dc527991feb248bae359741caca1/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L106)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

And ditto here on separating out the call to `httpGet()` into a separate line.
</blockquote></details>

Done.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 176 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-NAqrhqb5v4A_J15dpTk:-NAtMngr1uVtuwKERvDV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d21667e55805dc527991feb248bae359741caca1/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L176)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Since the other variable names for regexps all end with `RE`, how about making this one consistent with them?

```suggestion
	noRNRE                 = regexp.MustCompile(`[rR]elease [nN]ote: [nN]one`)
```
</blockquote></details>

Done.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 193 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-NAqtjqbDwjGo2Lx9w-R:-NAtREJGBZW0hJaE64Pl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d21667e55805dc527991feb248bae359741caca1/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L193)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

I'd suggest inserting `len(releaseNoteLines) > 0` directly into the `if` statements where `bufferNotEmpty` is used. The biggest reason is because possible side effects if the contents of `releaseNoteLines` change between the two uses of `bufferNotEmpty`. Using a pre-calculated value based on a variable that can change between the uses of the pre-calculated value in two `if` statements that are somewhat complex on their own is complicated. It's better to simplify and directly code in the calculation in both places.

I go back to these two ideas from the [Zen of Python](https://peps.python.org/pep-0020/) all the time. I think they apply here well:

> Simple is better than complex.
> Complex is better than complicated.
</blockquote></details>

Done.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 198 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-NAqzCCGEvlW2Aena4s3:-NAtRmat9NzXFkblrJ8R:b-cfwrsv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d21667e55805dc527991feb248bae359741caca1/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L198)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Does the comment for this `if` block state the same thing as the code itself states? If the words in the comment are more descriptive than the variable names, would it be better to change the names of the variables?

If you do choose to keep the comment, maybe update it to use language related to "release note lines" instead of the old "buffer" related language?
</blockquote></details>

I just omitted the comment as it largely does just restate what the code does.
___
*[`pkg/cmd/docs-issue-generation/docs_issue_generation.go` line 244 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-NAr0bwP5FpZ6ucl1Zyv:-NAtNL0RDt0mpmfRP_PL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d21667e55805dc527991feb248bae359741caca1/pkg/cmd/docs-issue-generation/docs_issue_generation.go#L244)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

It looks like this URL was hit by the search and replace too?

```suggestion
	postURL := "https://api.github.com/repos/cockroachdb/docs/issues"
 ```
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

jlinder reviewed (approved):
Excellent! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-:-NAtjUf1-pd7UVBACV6b:bmphau4)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jlinder and @nickvigilante)

___
*[`-- commits` line 14 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-NAqmYUh7fCKziq_9cvh):*
<details><summary><i>Previously, nickvigilante (Nick Vigilante) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

:thumbsup: 

This change could also be applied to the PR body.


<!-- Sent from Reviewable.io -->

nickvigilante reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85490)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jlinder)

___
*[`-- commits` line 14 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85490#-NAqmYUh7fCKziq_9cvh):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

:thumbsup: 

This change could also be applied to the PR body.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->
